README.md

```diff
@@ -67,14 +72,92 @@ final class Foo
+3. Repositories and Entities in public projects should not (and cannot) be defined as `final`. 
+4. Entity fields in public projects (vendors) should be `protected` instad of `private`.
+5. Decorate resource factories with decoration pattern and do not call resource instance with `new` keyword directly. Instead, inject resource factory into constructor and call `createNew()` on it. See `Sylius\Component\Product\Factory\ProductFactory`, `sylius.custom_factory.product` service definition and https://symfony.com/doc/current/service_container/service_decoration.html. The `priority` flag we are starting with equals 1 and is increased by one for each other decoration.
+6. For customizing forms use [Symfony Form Extension](https://symfony.com/doc/2.0/cookbook/form/create_form_type_extension.html).
+7. We follow command pattern implemented in [SyliusShopApiPlugin](https://github.com/Sylius/SyliusShopApiPlugin). This means we use the same bus libraries and similar `Command, CommandHandler, ViewRepository, ViewFactory, View` approach.
+
+
+### Testing
+
+0. Before you implement any new functional feature, write Behat scenario first (Gherkin, `*.feature` file).
+1. After writing the scenario, write a proper scenario execution (Contexts, Pages).
+2. Use Behat Contexts that are divided into `Hooks` - generic app Background, `Setup` specific resource background, `Ui` - specific interaction.
+3. Before starting implementing new functional code, make sure all your core logic is covered with PHPSpec (code without framework dependencies, like Commands, Forms, Configuration, Fixtures, etc.)
+4. PHPSpecs are always final classes with `function`s without `public` visibility and `: void` return type:
+
+```php
+final class ProductSpec extends ObjectBehavior
+{
+    function it_follows_bitbag_coding_standards(): void
+    {
+        Assert::true($this->followsStandards());
+    }
+}
+```
+
+
+## OOP / Architecture
+
+0. Make your code as simple as it's possible (follow single responsibilty principle).
+1. Use interfaces for any core logic class implementation, especially Models and Services (so that you follow single responsibilty principle).
+2. Use `final` any time it is possible (in order to avoid infinite inheritance chain, in order to customize some parts use Decorator and Dependency Injection patterns).
+3. Be careful with using static fields.
+4. Be more careful when you think Singleton is something you need in project. If it is you should go and rething the code.
+5. If that's not obvious yet, be careful with `static`, probably you will never need to use it.
+6. Don't use `Interface` suffix. By analysing the example below - It doesn't matter if it's an interface or an implementation. What matters is that `Foo` needs a cache, end of story. There's a principle behind interfaces: either you have one implementation and you don't need an interface, either you have many implementations and you create an interface.
```

Comment:
This one seems legit. What about dependency mocking with PHPSpec and vendor aware projects that could possibly use the decorator pattern for customization? In this case I think having an interface is relevant.

Reply:
Sure, but this rule does not forbid Intrefaces, I think it should be more like a warning or notice "Hey, is there any possibility to replace this implementation? Use different storage, queue, whatever? If no - you should think if you really need here an interface"